oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AccessTokenRequest.java

```diff
@@ -0,0 +1,139 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.auth.oauth2;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.auth.oauth2.ClientAuthentication;
+import com.linecorp.armeria.common.auth.oauth2.OAuth2Request;
+
+/**
+ * A OAuth 2.0 request to obtain an access token.
+ */
+@UnstableApi
+public interface AccessTokenRequest extends OAuth2Request {
+
+    /**
+     * Creates a new {@link AccessTokenRequest} that is used to request an access token using only the client
+     * credentials, as per
+     * <a href="https://datatracker.ietf.org/doc/html/rfc6749#section-4.4">RFC 6749, Section 4.4</a>.
+     */
+    static AccessTokenRequest ofClientCredentials(ClientAuthentication clientAuthentication) {
+        return ofClientCredentials(clientAuthentication, null);
+    }
+
+    /**
+     * Creates a new {@link AccessTokenRequest} that is used to request an access token using the client
+     * credentials and the scopes, as per
+     * <a href="https://datatracker.ietf.org/doc/html/rfc6749#section-4.4">RFC 6749, Section 4.4</a>.
+     */
+    static AccessTokenRequest ofClientCredentials(ClientAuthentication clientAuthentication,
+                                                  @Nullable List<String> scopes) {
+        requireNonNull(clientAuthentication, "clientAuthentication");
+        return new ClientCredentialsAccessTokenRequest(clientAuthentication, scopes);
+    }
+
+    /**
+     * Creates a new {@link AccessTokenRequest} that is used to request an access token using the client ID
+     * and client secret, as per
+     * <a href="https://datatracker.ietf.org/doc/html/rfc6749#section-4.4">RFC 6749, Section 4.4</a>.
+     * The specified client ID and client secret will be encoded and sent as the HTTP Basic authentication
+     * header.
+     */
+    static AccessTokenRequest ofClientCredentials(String clientId, String clientSecret) {
+        return ofClientCredentials(ClientAuthentication.ofClientPassword(clientId, clientSecret));
+    }
+
+    /**
+     * Creates a new {@link AccessTokenRequest} that is used to request an access token using the resource owner
+     * password credentials, as per
+     * <a href="https://datatracker.ietf.org/doc/html/rfc6749#section-4.3">RFC 6749, Section 4.3</a>.
+     */
+    static AccessTokenRequest ofResourceOwnerPassword(String username, String password) {
+        return ofResourceOwnerPassword(username, password, null, null);
```

Comment:
I left them nullable so that users can set a scope without ClientAuthentication` and I didn't want to create factory methods for all the number of cases.